### PR TITLE
Matthew getachews mini project

### PR DESCRIPTION
**Pull request by MatthewGetachew**
Creates a motor Talon that is able to be moved left or right with 2 xbox controller bumpers.

### CODE REVIEW

File: camp/mgetachew/mgetachewMiniProject/src/main/deploy/example.txt [hunk 1]
```diff
@@ -0,0 +1,3 @@
+Files placed in this directory will be deployed to the RoboRIO into the
```
Comment (josieammer) [inline camp/mgetachew/mgetachewMiniProject/src/main/deploy/example.txt]: remove unused file

File: camp/mgetachew/mgetachewMiniProject/src/main/java/frc/robot/Constants.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+ * constants. This class should not be used for any other purpose. All constants should be declared
+ * globally (i.e. public static). Do not put anything functional in this class.
+ *
+ * <p>It is advised to statically import this class (or one of its inner classes) wherever the
+ * constants are needed, to reduce verbosity.
+ */
+public final class Constants {
+  public static class OperatorConstants {
+    public static final int kDriverControllerPort = 0;
+    public static final double shooterOutfeedStatorCurrentMaximumAmps = 100.0;
```
Comment (josieammer) [inline camp/mgetachew/mgetachewMiniProject/src/main/java/frc/robot/Constants.java]: change the name of this variable. there isn't a shooterOutfeedMotor  in this project. Maybe just motorStatorCurrentMaximumAmps.

File: camp/mgetachew/mgetachewMiniProject/src/main/java/frc/robot/Constants.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    public static final double shooterOutfeedSupplyCurrentMaximumAmps = 50.0;
```
Comment (josieammer) [inline camp/mgetachew/mgetachewMiniProject/src/main/java/frc/robot/Constants.java]: same as above, change the name of this variable.

File: camp/mgetachew/mgetachewMiniProject/src/main/java/frc/robot/Constants.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    public static final double shooterOutfeedSupplyVoltageTimeConstant = 0.02;
+    //Motor Talon Port
+    public static final int motorTalonPort = 3;
+    //Change to fit your prefered xbox controller port
+    public static final int xboxControllerPort = 0;
+    //Could be related to the bug.
```
Comment (josieammer) [inline camp/mgetachew/mgetachewMiniProject/src/main/java/frc/robot/Constants.java]: take out unnecessary comment

File: camp/mgetachew/mgetachewMiniProject/src/main/java/frc/robot/Constants.java [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    public static InvertedValue rightTalonShooterMotorDefaultDirection = InvertedValue.CounterClockwise_Positive;  }
```
Comment (josieammer) [inline camp/mgetachew/mgetachewMiniProject/src/main/java/frc/robot/Constants.java]: rename this variable. there isn't a rightTalonShooter in this project.

File: camp/mgetachew/mgetachewMiniProject/src/main/java/frc/robot/RobotContainer.java [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+ * "declarative" paradigm, very little robot logic should actually be handled in the {@link Robot}
+ * periodic methods (other than the scheduler calls). Instead, the structure of the robot (including
+ * subsystems, commands, and trigger mappings) should be declared here.
+ */
+public class RobotContainer {
+  // The robot's subsystems and commands are defined here...
+  //Talon is set here
+  private final MotorTalon motorTalonSubsystem = new MotorTalon();
+
+  // Replace with CommandPS4Controller or CommandJoystick if needed
```
Comment (josieammer) [inline camp/mgetachew/mgetachewMiniProject/src/main/java/frc/robot/RobotContainer.java]: remove unnecessary comment

File: camp/mgetachew/mgetachewMiniProject/src/main/java/frc/robot/RobotContainer.java [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+   * Use this method to define your trigger->command mappings. Triggers can be created via the
+   * {@link Trigger#Trigger(java.util.function.BooleanSupplier)} constructor with an arbitrary
+   * predicate, or via the named factories in {@link
+   * edu.wpi.first.wpilibj2.command.button.CommandGenericHID}'s subclasses for {@link
+   * CommandXboxController Xbox}/{@link edu.wpi.first.wpilibj2.command.button.CommandPS4Controller
+   * PS4} controllers or {@link edu.wpi.first.wpilibj2.command.button.CommandJoystick Flight
+   * joysticks}.
+   */
+  private void configureBindings() {
+    // Schedule `ExampleCommand` when `exampleCondition` changes to `true`
```
Comment (josieammer) [inline camp/mgetachew/mgetachewMiniProject/src/main/java/frc/robot/RobotContainer.java]: take out unnecessary comments also below.

File: camp/mgetachew/mgetachewMiniProject/src/main/java/frc/robot/commands/Autos.java [hunk 8]
```diff
@@ -0,0 +1,20 @@
+// Copyright (c) FIRST and other WPILib contributors.
```
Comment (josieammer) [inline camp/mgetachew/mgetachewMiniProject/src/main/java/frc/robot/commands/Autos.java]: remove unnecessary file

File: camp/mgetachew/mgetachewMiniProject/src/main/java/frc/robot/commands/ExampleCommand.java [hunk 9]
```diff
@@ -0,0 +1,43 @@
+// Copyright (c) FIRST and other WPILib contributors.
```
Comment (josieammer) [inline camp/mgetachew/mgetachewMiniProject/src/main/java/frc/robot/commands/ExampleCommand.java]: remove unnecessary file

File: camp/mgetachew/mgetachewMiniProject/src/main/java/frc/robot/commands/moveMotor.java [hunk 10]
```diff
@@ -0,0 +1,43 @@
+package frc.robot.commands;
+
+import java.util.function.BooleanSupplier;
+
+import edu.wpi.first.wpilibj2.command.Command;
+import edu.wpi.first.wpilibj2.command.Subsystem;
+import frc.robot.subsystems.MotorTalon;
+
+public class moveMotor extends Command{
+    //globals here
```
Comment (josieammer) [inline camp/mgetachew/mgetachewMiniProject/src/main/java/frc/robot/commands/moveMotor.java]: remove unnecessary comment. They aren't globals, they are class variables.

File: camp/mgetachew/mgetachewMiniProject/src/main/java/frc/robot/commands/moveMotor.java [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+    MotorTalon motorTalonToMove;
+    BooleanSupplier rightBumperSupplier;
+    BooleanSupplier leftBumperSupplier;
+
+
+    public moveMotor(MotorTalon motorTalon, BooleanSupplier booleanXboxControllerSupplierLeft, BooleanSupplier booleanXboxControllerSupplierRight){
+        //require motor (or it wont work. Cant spin a motor without a motor)
```
Comment (josieammer) [inline camp/mgetachew/mgetachewMiniProject/src/main/java/frc/robot/commands/moveMotor.java]: remove unnecessary comment

File: camp/mgetachew/mgetachewMiniProject/src/main/java/frc/robot/commands/moveMotor.java [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+        addRequirements(motorTalon);
+        //construct the globals we set earlier
+        motorTalonToMove = motorTalon;
+        leftBumperSupplier = booleanXboxControllerSupplierLeft;
+        rightBumperSupplier = booleanXboxControllerSupplierRight;
+    }
+
+    @Override
+    public void execute(){
+        //left bumper is more importaint than the right bumper
```
Comment (josieammer) [inline camp/mgetachew/mgetachewMiniProject/src/main/java/frc/robot/commands/moveMotor.java]: would reword this to say:
when both bumpers are pressed, the left bumper is used.

File: camp/mgetachew/mgetachewMiniProject/src/main/java/frc/robot/commands/moveMotor.java [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+            motorTalonToMove.motorForwards();
+        } else {
+            //no bumper pressed stop the motor
+            motorTalonToMove.motorStop();
+        }
+    }
+
+    @Override
+    public boolean isFinished(){
+        //its never over
```
Comment (josieammer) [inline camp/mgetachew/mgetachewMiniProject/src/main/java/frc/robot/commands/moveMotor.java]: remove unnecessary comment

File: camp/mgetachew/mgetachewMiniProject/src/main/java/frc/robot/commands/moveMotor.java [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+    //globals here
+    MotorTalon motorTalonToMove;
+    BooleanSupplier rightBumperSupplier;
+    BooleanSupplier leftBumperSupplier;
+
+
+    public moveMotor(MotorTalon motorTalon, BooleanSupplier booleanXboxControllerSupplierLeft, BooleanSupplier booleanXboxControllerSupplierRight){
+        //require motor (or it wont work. Cant spin a motor without a motor)
+        addRequirements(motorTalon);
+        //construct the globals we set earlier
```
Comment (josieammer) [inline camp/mgetachew/mgetachewMiniProject/src/main/java/frc/robot/commands/moveMotor.java]: change comment. these aren't globals, they are class variables. We would say something like: "save constructor inputs to class variables"

File: camp/mgetachew/mgetachewMiniProject/src/main/java/frc/robot/subsystems/ExampleSubsystem.java [hunk 15]
```diff
@@ -0,0 +1,47 @@
+// Copyright (c) FIRST and other WPILib contributors.
```
Comment (josieammer) [inline camp/mgetachew/mgetachewMiniProject/src/main/java/frc/robot/subsystems/ExampleSubsystem.java]: remove unnecessary file

File: camp/mgetachew/mgetachewMiniProject/src/main/java/frc/robot/subsystems/MotorTalon.java [hunk 16]
```diff
@@ -0,0 +1,91 @@
+package frc.robot.subsystems;
+
+import edu.wpi.first.wpilibj.motorcontrol.PWMTalonFX;
```
Comment (josieammer) [inline camp/mgetachew/mgetachewMiniProject/src/main/java/frc/robot/subsystems/MotorTalon.java]: remove all unused imports

File: camp/mgetachew/mgetachewMiniProject/src/main/java/frc/robot/commands/moveMotor.java [hunk 17]
```diff
@@ -0,0 +1,43 @@
+package frc.robot.commands;
+
+import java.util.function.BooleanSupplier;
+
+import edu.wpi.first.wpilibj2.command.Command;
+import edu.wpi.first.wpilibj2.command.Subsystem;
```
Comment (josieammer) [inline camp/mgetachew/mgetachewMiniProject/src/main/java/frc/robot/commands/moveMotor.java]: remove unused import

File: camp/mgetachew/mgetachewMiniProject/src/main/java/frc/robot/subsystems/MotorTalon.java [hunk 18]
```diff
@@ -0,0 +1,91 @@
...
+public class MotorTalon extends SubsystemBase{
+    //configurtition for the motor begins here
+    private TalonFX motorTalon;
+    //velcity and voltage control
+    private final VelocityVoltage motorTalonVelocityController = new VelocityVoltage(0);
+    private final VoltageOut motorTalonVoltageController = new VoltageOut(0);
+    private TalonFXConfiguration motorTalonMotorConfiguration = null;
+    private Slot0Configs motorRpmGains = new Slot0Configs().withKP(0.36).withKI(0.1).withKD(0.0075).withKV(0.10);
+
+    private Slot0Configs motorTalonMotorHighRpmGains = new Slot0Configs().withKP(0.36).withKI(0.1).withKD(0.0075).withKV(0.10);
```
Comment (josieammer) [inline camp/mgetachew/mgetachewMiniProject/src/main/java/frc/robot/subsystems/MotorTalon.java]: remove unused code.

File: camp/mgetachew/mgetachewMiniProject/src/main/java/frc/robot/subsystems/MotorTalon.java [hunk 19]
```diff
@@ -0,0 +1,91 @@
...
+    private Slot1Configs motorTalonMotorLowRpmGains = new Slot1Configs().withKP(0.20).withKI(0.2).withKD(0.0075).withKV(0.05);
```
Comment (josieammer) [inline camp/mgetachew/mgetachewMiniProject/src/main/java/frc/robot/subsystems/MotorTalon.java]: remove unused code

File: camp/mgetachew/mgetachewMiniProject/src/main/java/frc/robot/subsystems/MotorTalon.java [hunk 20]
```diff
@@ -0,0 +1,91 @@
...
+
+    private NeutralModeValue outfeedMotorTargetNeutralModeValue = NeutralModeValue.Coast;
```
Comment (josieammer) [inline camp/mgetachew/mgetachewMiniProject/src/main/java/frc/robot/subsystems/MotorTalon.java]: change variable name. there is no outfeed motor in this project. maybe just motorTargetNeutralModeValue

File: camp/mgetachew/mgetachewMiniProject/src/main/java/frc/robot/subsystems/MotorTalon.java [hunk 21]
```diff
@@ -0,0 +1,91 @@
...
+    private static final double kMinDeadband = 0.001;
+      
+    //construct motor and print out the fact that its done.
+    public MotorTalon(){
+        motorTalon = new TalonFX(Constants.OperatorConstants.motorTalonPort);
+        configureMotor();
+        System.out.print("Motor is set");
+    }
+    //WARNING: Will go back and forth unpredictably
```
Comment (josieammer) [inline camp/mgetachew/mgetachewMiniProject/src/main/java/frc/robot/subsystems/MotorTalon.java]: remove unnecessary comment

File: camp/mgetachew/mgetachewMiniProject/src/main/java/frc/robot/subsystems/MotorTalon.java [hunk 22]
```diff
@@ -0,0 +1,91 @@
...
+    public void motorReverse(){
+        //reverse motor
+        motorTalon.setControl(this.motorTalonVelocityController.withVelocity(-1));
```
Comment (josieammer) [inline camp/mgetachew/mgetachewMiniProject/src/main/java/frc/robot/subsystems/MotorTalon.java]: -1 here (and 1 below) are magic numbers. Make a class variable that is the "reverseSpeedRpm" and "forwardSpeedRpm" and then use them here instead of -1 (and 1 below).

File: camp/mgetachew/mgetachewMiniProject/src/main/java/frc/robot/subsystems/MotorTalon.java [hunk 23]
```diff
@@ -0,0 +1,91 @@
...
+        //forwards motor
+        motorTalon.setControl(this.motorTalonVelocityController.withVelocity(1));
+    }
+
+    public void motorStop(){
+        //stop voltage
+        motorTalon.setControl(this.motorTalonVoltageController.withOutput(0));
+    }
+
+    //required. I wounldnt reccomend touching but if you know what your doing go ahead.
```
Comment (josieammer) [inline camp/mgetachew/mgetachewMiniProject/src/main/java/frc/robot/subsystems/MotorTalon.java]: remove unnecessary comment

File: camp/mgetachew/mgetachewMiniProject/src/main/java/frc/robot/subsystems/MotorTalon.java [hunk 24]
```diff
@@ -0,0 +1,91 @@
...
+import com.ctre.phoenix6.signals.NeutralModeValue;
+
+public class MotorTalon extends SubsystemBase{
+    //configurtition for the motor begins here
+    private TalonFX motorTalon;
+    //velcity and voltage control
+    private final VelocityVoltage motorTalonVelocityController = new VelocityVoltage(0);
+    private final VoltageOut motorTalonVoltageController = new VoltageOut(0);
+    private TalonFXConfiguration motorTalonMotorConfiguration = null;
+    private Slot0Configs motorRpmGains = new Slot0Configs().withKP(0.36).withKI(0.1).withKD(0.0075).withKV(0.10);
```
Comment (josieammer) [inline camp/mgetachew/mgetachewMiniProject/src/main/java/frc/robot/subsystems/MotorTalon.java]: these are not the right gains to use for this project. That's why it's not driving smoothly.
What you need here is:
Slot0Configs().withKS(0.5);

File: camp/mgetachew/mgetachewMiniProject/src/main/java/frc/robot/subsystems/MotorTalon.java [hunk 25]
```diff
@@ -0,0 +1,91 @@
...
+    private Slot1Configs motorTalonMotorLowRpmGains = new Slot1Configs().withKP(0.20).withKI(0.2).withKD(0.0075).withKV(0.05);
+
+    private NeutralModeValue outfeedMotorTargetNeutralModeValue = NeutralModeValue.Coast;
+    private static final double kMinDeadband = 0.001;
+      
+    //construct motor and print out the fact that its done.
+    public MotorTalon(){
+        motorTalon = new TalonFX(Constants.OperatorConstants.motorTalonPort);
+        configureMotor();
+        System.out.print("Motor is set");
```
Comment (josieammer) [inline camp/mgetachew/mgetachewMiniProject/src/main/java/frc/robot/subsystems/MotorTalon.java:34]: After you configure the motor, you need to have the Velocity Controller use Slot0. You do this with:
motorTalonVelocityControllerSlot = 0;

**Review by josieammer — CHANGES_REQUESTED** [review]
I left comments in your MotorTalonSubsystm to fix the bug you have with the stuttering motor. There is one line you have to change and one line you have to add.
There are a bunch of other code changes required. Please get these done as quickly as possible so that I have time to review before tomorrow's meeting. 
You will need to:
1) complete the requested code changes
2) re-test your code works on the board
3) merge in any changes from main that have been merged since you branched
4) push changes to the PR for final approval.

File: camp/mgetachew/mgetachewMiniProject/src/main/java/frc/robot/subsystems/MotorTalon.java (hunk 19 above)
Comment (MatthewGetachew) [reply camp/mgetachew/mgetachewMiniProject/src/main/java/frc/robot/subsystems/MotorTalon.java]: private Slot0Configs motorRpmGains = new Slot0Configs().withKP(0.36).withKI(0.1).withKD(0.0075).withKV(0.10);

is used by the config

**Comment by MatthewGetachew** [comment]
pushed the 2nd version with all changes applied. Just gotta test it

File: camp/mgetachew/mgetachewMiniProject/simgui-ds.json [hunk 26]
```diff
@@ -0,0 +1,92 @@
+{
```
Comment (josieammer) [inline camp/mgetachew/mgetachewMiniProject/simgui-ds.json]: remove unnecessary file

File: camp/mgetachew/mgetachewMiniProject/simgui.json [hunk 27]
```diff
@@ -0,0 +1,22 @@
+{
```
Comment (josieammer) [inline camp/mgetachew/mgetachewMiniProject/simgui.json]: remove unnecessary file

File: camp/mgetachew/mgetachewMiniProject/src/main/java/frc/robot/subsystems/MotorTalon.java [hunk 28]
```diff
@@ -0,0 +1,85 @@
...
+              "TalonFX ID " + motorTalon.getDeviceID() + " failed config with error " + response.toString());
+        }
+    
+        }
+    
+    
+
+    @Override
+    public void periodic() {
+        //print out the motor speed for debugging. can be commented out without any damage to the code.
```
Comment (josieammer) [inline camp/mgetachew/mgetachewMiniProject/src/main/java/frc/robot/subsystems/MotorTalon.java]: after you are done debugging, you need to remove this print statement. It spam's the logs.

File: camp/mgetachew/mgetachewMiniProject/src/main/java/frc/robot/subsystems/MotorTalon.java [hunk 29]
```diff
@@ -0,0 +1,85 @@
...
+      
+    //construct motor and print out the fact that its done.
+    public MotorTalon(){
+        motorTalon = new TalonFX(Constants.OperatorConstants.motorTalonPort);
+        configureMotor();
+        System.out.print("Motor is set");
+        motorTalonVelocityController.withSlot(0);
+    }
+
+    public void motorReverse(){
```
Comment (josieammer) [inline camp/mgetachew/mgetachewMiniProject/src/main/java/frc/robot/subsystems/MotorTalon.java]: all public methods must have a docstring. 
Type /** before this line to automatically generate the template.

Comment (josieammer) [inline camp/mgetachew/mgetachewMiniProject/src/main/java/frc/robot/subsystems/MotorTalon.java]: All methods must be in this order:

1) constructor(s)
2) public methods in alphabetical order
3) private methods in alphabetical order

File: camp/mgetachew/mgetachewMiniProject/src/main/java/frc/robot/subsystems/MotorTalon.java [hunk 30]
```diff
@@ -0,0 +1,85 @@
...
+        //reverse motor
+        motorTalon.setControl(this.motorTalonVelocityController.withVelocity(reverseSpeedRpm));
```
Comment (josieammer) [inline camp/mgetachew/mgetachewMiniProject/src/main/java/frc/robot/subsystems/MotorTalon.java]: I'm not sure the code works as written. Remember on the driver station laptop, we implemented a class variable, requestedSpeed, that was set by each of these motorReverse, motorForward, motorStop methods, and then in periodic, we set the value on the motor in an if-else statement. We also implemented a previousSpeed and tested whether the previous speed was the same as the requested speed and then only changed it on the motor if it was different. That may be required in order to fix the stutter. 
We can test this as-is and then add the code back if this doesn't work as you have written it here.

**Review by josieammer — CHANGES_REQUESTED** [review]
Some more minor changes. 
Also need to test on board and make sure it works as-written.

**Comment by MatthewGetachew** [comment]
finished

**Review by josieammer — APPROVED** [review]
Looks good to me!
